components/aws/sagemaker/common/sagemaker_component_spec.py

```diff
@@ -0,0 +1,198 @@
+SageMakerComponents."""
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+from dataclasses import dataclass
+from types import FunctionType
+
+from typing import (
+    Callable,
+    Generic,
+    List,
+    NamedTuple,
+    NewType,
+    TypeVar,
+    Dict,
+    Optional,
+    Union,
+    Any,
+)
+
+from .spec_validators import SpecValidators
+from .common_inputs import (
+    SageMakerComponentBaseInputs,
+    SageMakerComponentBaseOutputs,
+    SageMakerComponentInputValidator,
+    SageMakerComponentOutputValidator,
+    SageMakerIOValue,
+)
+
+IT = TypeVar("IT", bound=SageMakerComponentBaseInputs)  # Input Type
+OT = TypeVar("OT", bound=SageMakerComponentBaseOutputs)  # Output Type
+
+
+class SageMakerComponentSpec(Generic[IT, OT]):
+    """Defines the set of inputs and outputs as expected for a
+    SageMakerComponent.
+
+    This class represents the inputs and outputs that are required to be provided
+    to run a given SageMakerComponent. The component uses this to validate the
+    format of the input arguments as given by the pipeline at runtime. Components
+    should have a corresponding ComponentSpec inheriting from this
+    class and must override all private members:
+
+        INPUTS (as an inherited BaseInputs type)
+        OUTPUTS (as an inherited BaseOutputs type)
+
+    Typical usage example:
+
+        class MySageMakerComponentSpec(
+            SageMakerComponentSpec[SageMakerComponentInputs, SageMakerComponentOutputs]
+        ):
+            INPUTS = MySageMakerComponentInputs
+            OUTPUTS = MySageMakerComponentOutputs
+    """
+
+    # These inputs apply to all components
+    INPUTS: IT = SageMakerComponentBaseInputs()
+    OUTPUTS: OT = SageMakerComponentBaseOutputs()
+
+    OUTPUT_ARGUMENT_SUFFIX = "_output_path"
+
+    def __init__(
+        self,
+        arguments: List[str],
+        input_constructor: Callable[..., IT],
+        output_constructor: Callable[..., OT],
+    ):
+        """Instantiates the spec with given user inputs.
+
+        Args:
+            arguments: A list of command line arguments.
+            input_constructor: A constructor to create an input object.
+            output_constructor: A constructor to create an output object.
+        """
+        self._validate_spec()
+
+        parsed_args = self._parse_arguments(arguments)
+
+        # Split results into inputs and outputs
+        if self.INPUTS:
+            self._inputs: IT = input_constructor(
```

Comment:
whats going on with this method , where is it declared

Reply:
It's overridden by each component and will represent the constructor for that particular component's input dictionary type, as each component will have different inputs.